File: android-native/app/src/main/java/one/zrp/social/mobile/ui/navigation/UnreadBadgeViewModel.kt
package one.zrp.social.mobile.ui.navigation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import one.zrp.social.mobile.data.MessagesRepository
import one.zrp.social.mobile.data.NotificationsRepository
import one.zrp.social.mobile.util.aggregateUnreadCount

/**
 * Backs the bottom nav's unread-notifications badge - real
 * GET /notifications/unread, the same endpoint the website's bell icon
 * uses for its own unread dot. Scoped above the NavHost (created once,
 * in ZrpNavHost, not per-screen) since the badge needs to stay visible
 * while any tab is showing, not just the Notifications tab's own
 * ViewModel, which only exists while that screen is composed.
 */
class UnreadBadgeViewModel(private val repository: NotificationsRepository) : ViewModel() {
    private val _unreadCount = MutableStateFlow(0)
    val unreadCount: StateFlow<Int> = _unreadCount.asStateFlow()

    init {
        refresh()
    }

    fun refresh() {
        viewModelScope.launch {
            repository.getUnreadCount().onSuccess { count -> _unreadCount.value = count }
        }
    }

    // Called the moment the Notifications tab is selected - the real
    // GET /notifications/unread count won't reflect the mark-all-read
    // call NotificationsViewModel makes until that request round-trips,
    // so the badge clears immediately here rather than showing a stale
    // count for that gap, then a follow-up refresh() (fired on the next
    // tab switch away) confirms the server agrees.
    fun clear() {
        _unreadCount.value = 0
    }
}

class UnreadBadgeViewModelFactory(private val repository: NotificationsRepository) : ViewModelProvider.Factory {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return UnreadBadgeViewModel(repository) as T
    }
}

/**
 * Backs the bottom nav's unread-messages badge - real GET
 * /messages/unread (1:1) PLUS every real GROUP conversation's own
 * unreadCount (GET /conversations - see aggregateUnreadCount's own
 * KDoc for why the two have to be summed client-side), mirroring
 * [UnreadBadgeViewModel] above for notifications. No clear(): unlike
 * Notifications, opening the Messages tab (the conversation list)
 * doesn't itself mark anything read on web either - only opening a
 * specific conversation does (ConversationViewModel/
 * GroupConversationViewModel's own read-marking) - so there's no
 * "visiting this tab just cleared everything" moment to reflect
 * instantly the way Notifications has.
 */
class UnreadMessagesBadgeViewModel(private val repository: MessagesRepository) : ViewModel() {
    private val _unreadCount = MutableStateFlow(0)
    val unreadCount: StateFlow<Int> = _unreadCount.asStateFlow()

    init {
        refresh()
    }

    fun refresh() {
        viewModelScope.launch {
            val directCount = repository.getUnreadCount().getOrDefault(0)
            val groupConversations = repository.getGroupConversations().getOrDefault(emptyList())
            _unreadCount.value = aggregateUnreadCount(directCount, groupConversations)
        }
    }
}

class UnreadMessagesBadgeViewModelFactory(private val repository: MessagesRepository) : ViewModelProvider.Factory {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return UnreadMessagesBadgeViewModel(repository) as T
    }
}

File: ios-native/ZRPSocial/Features/Messages/UnreadBadgeViewModel.swift
import Foundation
import SwiftUI

/// The unread counts behind the Messages and Notifications badges.
///
/// Reads `GET /api/messages/unread` - the same dedicated endpoint the
/// website's own badge uses - rather than deriving a number from an
/// already-fetched conversation list, which the toolbar does not have
/// access to and which would be stale the moment a message arrived.
@MainActor
final class UnreadBadgeViewModel: ObservableObject {

    @Published private(set) var messageCount = 0
    @Published private(set) var notificationCount = 0

    private let repository: MessagesRepositoryProtocol
    private let notifications: NotificationsRepositoryProtocol

    init(
        repository: MessagesRepositoryProtocol = MessagesRepository(),
        notifications: NotificationsRepositoryProtocol = NotificationsRepository()
    ) {
        self.repository = repository
        self.notifications = notifications
    }

    /// A failure here is deliberately silent: an unread badge is not
    /// worth an error state, and the count simply stays as it was.
    func refresh() async {
        // Two independent counts from two dedicated endpoints, fetched
        // together so one failing does not hide the other.
        async let messages = try? repository.unreadCount()
        async let alerts = try? notifications.unreadCount()
        let (messageResult, alertResult) = await (messages, alerts)
        if let messageResult { messageCount = messageResult }
        if let alertResult { notificationCount = alertResult }
    }

    /// Called after the notifications list marks everything read, so the
    /// badge clears without waiting for the next refresh.
    func clearNotificationCount() {
        notificationCount = 0
    }
}
